Automattic/jetpack
# CRM: Fixing segments by email and doesnotcontain, by reordering checks in segmentConditionArgs()

## Proposed changes:

* When adding new segments using the 'Does Not Contain' condition in a few situations (using Email, or Address fields), an error popup will show and the segment won't save. This PR addresses the issue with the Email field by changing the order in which condition types are checked in `segmentConditionArgs`.
* This PR also changes the `segment_condition_produces_no_args` error code to produce a 400 response instead of 500, when previewing segments.

### Other information:


## Jetpack product discussion

https://github.com/Automattic/jetpack-crm-extensions/issues/672
Related PR: https://github.com/Automattic/jetpack-crm-extensions/pull/675

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

1. To test this out, install and activate the Advanced Segments extension on a test site, as well as having the CRM plugin installed and active.
2. Open up the Segments dashboard, and visit the extensions page: [`/wp-admin/admin.php?page=manage-segments`](http://norubbish.local/wp-admin/admin.php?page=manage-segments), then click 'Add New'.
3. Attempt to create the following segment:
   1. Email / Does not contain / {example-email-string}
4. Each time you click 'Preview Segment', you will see a popup with 'General Error `#219` There was a general error' on it
5. Each time you click 'Save', the page will display the following message:

<img width="625" alt="235457834-71b07c6f-0807-4b90-92ce-a67191c09d3f" src="https://user-images.githubusercontent.com/16754605/235901893-0b14c8b6-9864-4e1b-a54e-7923f31512f2.png">

6. Apply this PR, and follow the same steps - the segments should save properly and a preview be visible as well.
7. To test the error code change, the best bet is to apply this PR but not the Advanced Segments PR (ideally done before that PR is merged - otherwise comment out the changed lines in `contact/class-segment-condition-contact-address-field.php`).
8. Then, try adding a 'Does Not Contain' condition to an address field when building a segment. Click 'Preview' and check the developer console - you should see `POST .../wp-admin/admin-ajax.php 400 (Bad Request)` instead of `POST .../wp-admin/admin-ajax.php 500 (Internal Server Error)`
